pkg/bigquerysql/ddl.go

```diff
@@ -0,0 +1,133 @@
+			log.Warn("BigQuery does not support update column nullable", zap.String("column", diff.After.Name), zap.Any("before", diff.Before.Nullable), zap.Any("after", diff.After.Nullable))
+		} else {
+			strs = append(strs, fmt.Sprintf("COLUMN %s SET NOT NULL", diff.After.Name))
+		}
+	}
+	return strings.Join(strs, ", "), nil
+}
+
+func GenDDLViaColumnsDiff(datasetID, tableID string, prevColumns []cloudstorage.TableCol, curTableDef cloudstorage.TableDefinition) ([]string, error) {
+	tableFullName := fmt.Sprintf("%s.%s", datasetID, tableID)
+
+	if curTableDef.Type == timodel.ActionTruncateTable {
+		return []string{fmt.Sprintf("TRUNCATE TABLE %s", tableFullName)}, nil
+	}
+	if curTableDef.Type == timodel.ActionDropTable {
+		return []string{fmt.Sprintf("DROP TABLE %s", tableFullName)}, nil
+	}
+	if curTableDef.Type == timodel.ActionCreateTable {
+		return nil, errors.New("Received create table ddl, which should not happen") // FIXME: drop table and create table
+	}
+	if curTableDef.Type == timodel.ActionRenameTables {
+		return nil, errors.New("Received rename table ddl, new change data can not be capture by TiCDC any more." +
+			"If you want to rename table, please start a new task to capture the new table") // FIXME: rename table to new table and rename back
+	}
+	if curTableDef.Type == timodel.ActionDropSchema {
+		return nil, errors.New("Received drop schema ddl, which does not support") // FIXME: drop schema and create schema
+	}
+	if curTableDef.Type == timodel.ActionCreateSchema {
+		return nil, errors.New("Received create schema ddl, which should not happen") // FIXME: drop schema and create schema
+	}
+
+	columnDiff, err := tidbsql.GetColumnDiff(prevColumns, curTableDef.Columns)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	ddls := make([]string, 0, len(columnDiff))
+	for _, item := range columnDiff {
+		ddl := ""
+		switch item.Action {
+		case tidbsql.ADD_COLUMN:
+			ddl += fmt.Sprintf("ALTER TABLE %s ADD COLUMN ", tableFullName)
+			colStr, err := GetBigQueryColumnString(*item.After)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ddl += colStr
+		case tidbsql.DROP_COLUMN:
+			ddl += fmt.Sprintf("ALTER TABLE %s DROP COLUMN %s", tableFullName, item.Before.Name)
+		case tidbsql.MODIFY_COLUMN:
+			ddl += fmt.Sprintf("ALTER TABLE %s ALTER COLUMN ", tableFullName)
+			modifyStr, err := GetColumnModifyString(&item)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ddl += modifyStr
+		case tidbsql.RENAME_COLUMN:
+			ddl += fmt.Sprintf("ALTER TABLE %s RENAME COLUMN %s TO %s", tableFullName, item.Before.Name, item.After.Name)
+		default:
+			// UNCHANGE
+		}
+		if ddl != "" {
+			ddl += ";"
+			ddls = append(ddls, ddl)
+		}
+	}
+
+	// TODO: handle primary key
+	return ddls, nil
+}
+
+func getDefaultString(val interface{}) string {
+	_, err := strconv.ParseFloat(fmt.Sprintf("%v", val), 64)
+	if err != nil {
+		return fmt.Sprintf("'%v'", val) // FIXME: escape
+	}
+	return fmt.Sprintf("%v", val)
+}
```

Comment:
Can bigquery handle `alter table t add column a int default "1"`? If yes, this function is not needed.

Reply:
When setting the default value of a string column to 'a', bigquery can not handle `alter table t alter column str_coulmn set default a`, so this function is still necessary.